src/mlpack/bindings/python/print_input_processing.hpp

```diff
@@ -54,35 +55,132 @@ void PrintInputProcessing(
    *
    * # Detect if the parameter was passed; set if so.
    * if param_name is not None:
+   *   if isinstance(param_name, int):
    *   SetParam[int](<const string> 'param_name', param_name)
    *   CLI.SetPassed(<const string> 'param_name')
+   * else:
+   *   raise TypeError("'param_name' must have type 'list'!")
    */
+
+
   std::cout << prefix << "# Detect if the parameter was passed; set if so."
       << std::endl;
   if (!d.required)
   {
-    std::cout << prefix << "if " << name << " is not " << def << ":"
+    if (def == "False")
+    {
+      std::cout << prefix << "if isinstance(" << name << ", bool):"
+        << std::endl;
+      std::cout << prefix << "  if " << name << " is not " << def << ":"
+        << std::endl;
+    }
+    else
+    {
+      std::cout << prefix << "if " << name << " is not " << def << ":"
+        << std::endl;
+    }
+    if (GetCythonType<T>(d) == "string")
+    {
+      std::cout << prefix << "  if isinstance(" << name << ", str):"
         << std::endl;
+    }
+    else if (GetCythonType<T>(d) == "vector[string]" ||
+        GetCythonType<T>(d) == "vector[int]" ||
+        GetCythonType<T>(d) == "vector[double]")
+    {
+      std::cout << prefix << "  if isinstance(" << name << ", list):"
+        << std::endl;
+    }
+    else if (GetCythonType<T>(d) == "double")
+    {
+      std::cout << prefix << "  if isinstance(" << name << ", float):"
+        << std::endl;
+    }
+    else if (GetCythonType<T>(d) == "int")
+    {
+      std::cout << prefix << "  if isinstance(" << name << ", int):"
+        << std::endl;
+    }
```

Comment:
Actually I think you can use `GetPrintableType()` here... take a look (`get_printable_type.hpp`) and tell me what you think.

```
std::cout << prefix << "  if isinstance(" << name << ", " << GetPrintableType<T>(d)
    << "):" << std::endl;
```

That could simplify this code quite a lot.

Reply:
If we use `GetPrintableType()` at the place of `int` or `float` or `str` it is not recognized by python as there is not datatype name `double` or `string`

Reply:
That's a fair point.  The `GetPrintableType<>()` functions are used solely for printing documentation, so I would be okay adapting the typenames returned by that function to be the correct Python type.  Specifically I can see in the code that `double` should change to `float` and `string` to `str`.  But this still does not fully solve the problem for `list of ...` and the matrix types.

I think that for list types, we should make another overload of this function `PrintInputProcessing()` that is specifically for vector types.  This would mean using SFINAE to make the following function:

```
template<typename T>
void PrintInputProcessing(
    const util::ParamData& d,
    const size_t indent,
    const typename boost::enable_if<util::IsStdVector<T>::type* = 0,
    const typename boost::disable_if<arma::is_arma_type<T>>::type* = 0,
    const typename boost::disable_if<data::HasSerialize<T>>::type* = 0,
    const typename boost::disable_if<std::is_same<T,
        std::tuple<data::DatasetInfo, arma::mat>>>::type* = 0)
```

and the implementation of that function would be basically the same as the one I am commenting on here, except that we would need to test two things: `isinstance(name, list)` and, if the length of the list is greater than 0, `isinstance(name[0], PT)` where `PT` is the Python type of the list element (you could access this type in C++ as `T::value_type`, so I think you could call `GetPrintableType<T::value_type>()` to get the right result here.

In addition, because of the SFINAE signature, you'll have to add the following parameter to all of the other `PrintInputProcessing()` overloads (except the one at the bottom of the file):

```
const typename boost::disable_if<util::IsStdVector<T>>::type* = 0
```

Do let me know if I can clarify any of this.  SFINAE code can be ugly and confusing...